File: test/operator/test_ttno_graph.c
#include "ttno_graph.h"
#include "ttno.h"
#include "aligned_memory.h"


#define ARRLEN(a) (sizeof(a) / sizeof(a[0]))


char* test_ttno_graph_from_opchains()
{
	hid_t file = H5Fopen("../test/operator/data/test_ttno_graph_from_opchains.hdf5", H5F_ACC_RDONLY, H5P_DEFAULT);
	if (file < 0) {
		return "'H5Fopen' in test_ttno_graph_from_opchains failed";
	}

	// number of physical and branching lattice sites
	const int nsites_physical  = 5;
	const int nsites_branching = 3;
	const int nsites = nsites_physical + nsites_branching;
	// local physical dimension
	const long d = 3;

	const long dim_full_physical = ipow(d, nsites_physical);

	// tree topology:
	//
	//     4     0
	//      ╲   ╱
	//       ╲ ╱
	//        6
	//        │
	//        │
	//  2 ─── 5 ─── 1 ─── 7
	//        │
	//        │
	//        3
	//
	int neigh0[] = { 6 };
	int neigh1[] = { 5, 7 };
	int neigh2[] = { 5 };
	int neigh3[] = { 5 };
	int neigh4[] = { 6 };
	int neigh5[] = { 1, 2, 3, 6 };
	int neigh6[] = { 0, 4, 5 };
	int neigh7[] = { 1 };
	int* neighbor_map[8] = {
		neigh0, neigh1, neigh2, neigh3, neigh4, neigh5, neigh6, neigh7,
	};
	int num_neighbors[8] = {
		ARRLEN(neigh0), ARRLEN(neigh1), ARRLEN(neigh2), ARRLEN(neigh3), ARRLEN(neigh4), ARRLEN(neigh5), ARRLEN(neigh6), ARRLEN(neigh7),
	};
	struct abstract_graph topology = {
		.neighbor_map  = neighbor_map,
		.num_neighbors = num_neighbors,
		.num_nodes     = nsites,
	};
	assert(abstract_graph_is_connected_tree(&topology));

	const int nchains = 9;
	struct op_chain* chains = ct_malloc(nchains * sizeof(struct op_chain));
	for (int i = 0; i < nchains; i++)
	{
		char varname[1024];
		sprintf(varname, "/chain%i/length", i);
		int length;
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, &length) < 0) {
			return "reading operator chain length from disk failed";
		}
		allocate_op_chain(length, &chains[i]);

		sprintf(varname, "/chain%i/oids", i);
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, chains[i].oids) < 0) {
			return "reading operator chain IDs from disk failed";
		}
		sprintf(varname, "/chain%i/qnums", i);
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, chains[i].qnums) < 0) {
			return "reading operator chain quantum numbers from disk failed";
		}
		sprintf(varname, "/chain%i/cid", i);
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, &chains[i].cid) < 0) {
			return "reading operator chain coefficient index from disk failed";
		}
		sprintf(varname, "/chain%i/istart", i);
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, &chains[i].istart) < 0) {
			return "reading operator chain starting site index from disk failed";
		}
	}

	struct ttno_graph graph;
	if (ttno_graph_from_opchains(chains, nchains, nsites_physical, &topology, &graph) < 0) {
		return "'ttno_graph_from_opchains' failed internally";
	}
	if (!ttno_graph_is_consistent(&graph)) {
		return "constructed TTNO graph is inconsistent";
	}

	// bond dimensions
	int rank_04_123;
	if (read_hdf5_attribute(file, "rank_04_123", H5T_NATIVE_INT, &rank_04_123) < 0) {
		return "reading operator partitioning rank from disk failed";
	}
	int rank_1_0234;
	if (read_hdf5_attribute(file, "rank_1_0234", H5T_NATIVE_INT, &rank_1_0234) < 0) {
		return "reading operator partitioning rank from disk failed";
	}
	int rank_2_0134;
	if (read_hdf5_attribute(file, "rank_2_0134", H5T_NATIVE_INT, &rank_2_0134) < 0) {
		return "reading operator partitioning rank from disk failed";
	}
	int rank_4_0123;
	if (read_hdf5_attribute(file, "rank_4_0123", H5T_NATIVE_INT, &rank_4_0123) < 0) {
		return "reading operator partitioning rank from disk failed";
	}

	if (graph.num_verts[5*nsites + 6] != rank_04_123) {
		return "virtual bond dimension does not match expected partitioning matrix rank";
	}
	if (graph.num_verts[1*nsites + 5] != rank_1_0234) {
		return "virtual bond dimension does not match expected partitioning matrix rank";
	}
	if (graph.num_verts[2*nsites + 5] != rank_2_0134) {
		return "virtual bond dimension does not match expected partitioning matrix rank";
	}
	if (graph.num_verts[4*nsites + 6] != rank_4_0123) {
		return "virtual bond dimension does not match expected partitioning matrix rank";
	}

	// read local operator map from disk
	const int num_local_ops = 18;
	struct dense_tensor opmap_tensor;
	const long dim_opmt[3] = { num_local_ops, d, d };
	allocate_dense_tensor(CT_DOUBLE_COMPLEX, 3, dim_opmt, &opmap_tensor);
	// read values from disk
	if (read_hdf5_dataset(file, "opmap", H5T_NATIVE_DOUBLE, opmap_tensor.data) < 0) {
		return "reading tensor entries from disk failed";
	}
	// copy individual operators
	struct dense_tensor* opmap = ct_malloc(num_local_ops * sizeof(struct dense_tensor));
	for (int i = 0; i < num_local_ops; i++)
	{
		const long dim[2] = { d, d };
		allocate_dense_tensor(CT_DOUBLE_COMPLEX, 2, dim, &opmap[i]);
		const dcomplex* data = opmap_tensor.data;
		memcpy(opmap[i].data, &data[i * d*d], d*d * sizeof(dcomplex));
	}
	delete_dense_tensor(&opmap_tensor);

	// coefficient map
	dcomplex* coeffmap = ct_malloc(9 * sizeof(dcomplex));
	if (read_hdf5_dataset(file, "coeffmap", H5T_NATIVE_DOUBLE, coeffmap) < 0) {
		return "reading coefficient map from disk failed";
	}

	// convert TTNO graph to a (dense) matrix
	struct dense_tensor mat;
	ttno_graph_to_matrix(&graph, opmap, coeffmap, &mat);

	// sum matrix representations of individual operator chains, as reference
	struct dense_tensor mat_ref;
	const long dim_mat_ref[2] = { dim_full_physical, dim_full_physical };
	allocate_dense_tensor(CT_DOUBLE_COMPLEX, 2, dim_mat_ref, &mat_ref);
	for (int i = 0; i < nchains; i++)
	{
		struct dense_tensor c;
		op_chain_to_matrix(&chains[i], d, nsites_physical, opmap, coeffmap, CT_DOUBLE_COMPLEX, &c);
		dense_tensor_scalar_multiply_add(numeric_one(CT_DOUBLE_COMPLEX), &c, &mat_ref);
		delete_dense_tensor(&c);
	}

	// compare
	if (!dense_tensor_allclose(&mat, &mat_ref, 1e-13)) {
		return "matrix representation of TTNO graph does not match reference matrix";
	}

	// clean up
	delete_dense_tensor(&mat_ref);
	delete_dense_tensor(&mat);
	ct_free(coeffmap);
	for (int i = 0; i < num_local_ops; i++) {
		delete_dense_tensor(&opmap[i]);
	}
	ct_free(opmap);
	for (int i = 0; i < nchains; i++)
	{
		delete_op_chain(&chains[i]);
	}
	ct_free(chains);
	delete_ttno_graph(&graph);

	H5Fclose(file);

	return 0;
}

File: test/operator/test_ttno_graph.py
import numpy as np
import h5py
import pytenet as ptn


def ttno_graph_from_opchains_data():

    # random number generator
    rng = np.random.default_rng(307)

    # physical quantum numbers
    qd = np.array([0, -1, 1])
    # local physical dimension
    d = len(qd)
    # number of physical sites
    nsites_physical = 5

    # identity operator ID
    oid_identity = 0
    # number of local operators
    num_local_ops = 18

    coeffmap = np.concatenate((np.array([0., 1.]), ptn.crandn(7, rng)))

    cids = [5, 8, 7, 4, 3, 2, 3, 6, 4]
    chains = [  #     0   1   2   3   4       0   1   2   3   4
        ptn.OpChain([     6,  9,  5, 15], [     0, -1, -1,  1,  0], coeffmap[cids[0]], 1),
        ptn.OpChain([ 8, 16,  2, 11,  3], [ 0, -1,  1,  2,  1,  0], coeffmap[cids[1]], 0),
        ptn.OpChain([        10,  1    ], [         0,  1,  0    ], coeffmap[cids[2]], 2),
        ptn.OpChain([ 9,  7, 13, 12,  1], [ 0,  0, -1, -1,  1,  0], coeffmap[cids[3]], 0),
        ptn.OpChain([     0, 17,  4,  7], [     0,  0,  1,  1,  0], coeffmap[cids[4]], 1),
        ptn.OpChain([            15, 14], [             0, -1,  0], coeffmap[cids[5]], 3),
        ptn.OpChain([     9,  2,  1    ], [     0,  0,  1,  0    ], coeffmap[cids[6]], 1),
        ptn.OpChain([ 7,  0,  0, 14,  9], [ 0, -1, -1, -1,  0,  0], coeffmap[cids[7]], 0),
        ptn.OpChain([        10,  0,  7], [         0,  1,  1,  0], coeffmap[cids[8]], 2),
    ]

    graph = ptn.OpGraph.from_opchains(chains, nsites_physical, oid_identity)
    assert graph.is_consistent()
    assert graph.length == nsites_physical

    # random local operators
    opmap = [np.identity(len(qd), dtype=complex) if opid == oid_identity else ptn.crandn((d, d), rng)
             for opid in range(num_local_ops)]
    # enforce sparsity pattern according to quantum numbers
    for chain in chains:
        for i, opid in enumerate(chain.oids):
            qDloc = chain.qnums[i:i+2]
            mask = ptn.qnumber_outer_sum([qd, -qd, [qDloc[0]], [-qDloc[1]]])[:, :, 0, 0]
            opmap[opid] = np.where(mask == 0, opmap[opid], 0)
    # sparsity pattern should not lead to zero operators
    for op in opmap:
        assert np.linalg.norm(op) > 0

    # reference matrix representation of operator chains
    mat_ref = 0
    for chain in chains:
        # including leading and trailing identity maps
        mat_ref = mat_ref + np.kron(np.kron(
            np.identity(len(qd)**chain.istart),
            chain.as_matrix(opmap)),
            np.identity(len(qd)**(nsites_physical - (chain.istart + chain.length))))

    # group sites (0, 4) and (1, 2, 3)
    rank_04_123 = _operator_partition_rank(mat_ref, d, (0, 4), (1, 2, 3))
    # group sites (1,) and (0, 2, 3, 4)
    rank_1_0234 = _operator_partition_rank(mat_ref, d, (1,), (0, 2, 3, 4))
    # group sites (2) and (0, 1, 3, 4)
    rank_2_0134 = _operator_partition_rank(mat_ref, d, (2,), (0, 1, 3, 4))
    # group sites (4) and (0, 1, 2, 3)
    rank_4_0123 = _operator_partition_rank(mat_ref, d, (4,), (0, 1, 2, 3))

    with h5py.File("data/test_ttno_graph_from_opchains.hdf5", "w") as file:
        for i, chain in enumerate(chains):
            file.attrs[f"/chain{i}/length"] = chain.length
            file.attrs[f"/chain{i}/oids"]   = chain.oids
            file.attrs[f"/chain{i}/qnums"]  = chain.qnums
            file.attrs[f"/chain{i}/cid"]    = cids[i]
            file.attrs[f"/chain{i}/istart"] = chain.istart
        file["opmap"]    = np.array(opmap)
        file["coeffmap"] = coeffmap
        file.attrs["rank_04_123"] = rank_04_123
        file.attrs["rank_1_0234"] = rank_1_0234
        # note: local operators acting on sites 2 and 4 span a 5-dimensional subspace,
        # hence numerically determined rank is smaller than general abstract case
        file.attrs["rank_2_0134"] = rank_2_0134 + 1
        file.attrs["rank_4_0123"] = rank_4_0123 + 2


def _operator_partition_rank(u, d: int, sites_a, sites_b):
    """
    Compute the matrix rank when partitioning an operator
    between two subsystems corresponding to 'sites_a' and 'sites_b'.
    """
    sites_a = tuple(sites_a)
    sites_b = tuple(sites_b)
    assert set(sites_a).isdisjoint(sites_b)
    na = len(sites_a)
    nb = len(sites_b)
    u = _permute_operation(u, d, sites_a + sites_b)
    u = u.reshape((d**na, d**nb, d**na, d**nb))
    u = np.transpose(u, (0, 2, 1, 3))
    u = u.reshape((d**(2*na), d**(2*nb)))
    return np.linalg.matrix_rank(u)


def _permute_operation(u: np.ndarray, d: int, perm):
    """
    Find the representation of a matrix after permuting lattice sites.
    """
    nsites = len(perm)
    assert u.shape == (d**nsites, d**nsites)
    perm = list(perm)
    u = np.reshape(u, (2*nsites) * (d,))
    u = np.transpose(u, perm + [nsites + p for p in perm])
    u = np.reshape(u, (d**nsites, d**nsites))
    return u


def main():
    ttno_graph_from_opchains_data()


if __name__ == "__main__":
    main()
